File: backend/app/api/drills.py
from typing import Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.db.deps import get_db
from app.schemas.drill import (
    DrillCreate, DrillRead, DrillUpdate,
    DrillCommentCreate, DrillCommentRead
)
from app.services.drill_service import DrillService
from app.services.drill_comment_service import DrillCommentService
from app.schemas.drill import DrillRatingCreate, DrillRatingRead

from app.services.drill_rating_service import DrillRatingService
router = APIRouter(prefix="/drills", tags=["Drills"])


@router.post("", response_model=DrillRead)
def create_drill(payload: DrillCreate, db: Session = Depends(get_db)):
    service = DrillService(db)
    return service.create_drill(payload)


@router.get("", response_model=list[DrillRead])
def list_drills(
    levels: Optional[str] = Query(None),
    fundamentals: Optional[str] = Query(None),
    drill_types: Optional[str] = Query(None),
    q: Optional[str] = Query(None),
    coach: Optional[bool] = Query(None),
    many: Optional[bool] = Query(None),
    db: Session = Depends(get_db),
):
    service = DrillService(db)
    return service.list_drills(
        levels=levels,
        fundamentals=fundamentals,
        drill_types=drill_types,
        q=q,
        coach=coach,
        many=many,
    )


# ✅ CRUD
@router.get("/{drill_id}", response_model=DrillRead)
def get_drill(drill_id: str, db: Session = Depends(get_db)):
    service = DrillService(db)
    return service.get_drill(drill_id)


@router.put("/{drill_id}", response_model=DrillRead)
def update_drill(drill_id: str, payload: DrillUpdate, db: Session = Depends(get_db)):
    service = DrillService(db)
    return service.update_drill(drill_id, payload)


@router.delete("/{drill_id}")
def delete_drill(drill_id: str, db: Session = Depends(get_db)):
    service = DrillService(db)
    service.delete_drill(drill_id)
    return {"ok": True}


# ✅ Comments
@router.get("/{drill_id}/comments", response_model=list[DrillCommentRead])
def list_comments(drill_id: str, db: Session = Depends(get_db)):
    service = DrillCommentService(db)
    return service.list_comments(drill_id)


@router.post("/{drill_id}/comments", response_model=DrillCommentRead)
def add_comment(drill_id: str, payload: DrillCommentCreate, db: Session = Depends(get_db)):
    service = DrillCommentService(db)
    return service.add_comment(drill_id, payload)


@router.delete("/comments/{comment_id}")
def delete_comment(comment_id: str, db: Session = Depends(get_db)):
    service = DrillCommentService(db)
    service.delete_comment(comment_id)
    return {"ok": True}

@router.get("/{drill_id}/ratings", response_model=list[DrillRatingRead])
def list_ratings(drill_id: str, db: Session = Depends(get_db)):
    service = DrillRatingService(db)
    return service.list_ratings(drill_id)

@router.post("/{drill_id}/ratings", response_model=DrillRatingRead)
def add_rating(drill_id: str, payload: DrillRatingCreate, db: Session = Depends(get_db)):
    service = DrillRatingService(db)
    return service.add_rating(drill_id, payload)

@router.delete("/ratings/{rating_id}")
def delete_rating(rating_id: str, db: Session = Depends(get_db)):
    service = DrillRatingService(db)
    service.delete_rating(rating_id)
    return {"ok": True}

File: frontend/src/api/drills.js
import { supabase } from "../lib/supabaseClient";

function applyFilters(query, params) {
  const { levels, fundamentals, drill_types, coach, many } = params || {};

  if (levels?.length) query = query.overlaps("levels", levels);
  if (fundamentals?.length) query = query.overlaps("fundamentals", fundamentals);
  if (drill_types?.length) query = query.overlaps("drill_types", drill_types);

  if (coach === true) query = query.eq("coach_participates", true);
  if (many === true) query = query.eq("good_for_many_players", true);

  return query;
}

async function requireAuth(msg = "Login required.") {
  const { data, error } = await supabase.auth.getUser();
  if (error) throw new Error(error.message);
  if (!data?.user) throw new Error(msg);
  return data.user;
}

// ---------- Drills (public read via view) ----------
export async function listDrills(params = {}) {
  let q = supabase.from("drills_public").select("*");
  q = applyFilters(q, params);

  q = q
    .order("has_rating", { ascending: false })
    .order("avg_rating", { ascending: false, nullsFirst: false })
    .order("ratings_count", { ascending: false })
    .order("has_comments", { ascending: false })
    .order("comments_count", { ascending: false })
    .order("created_at", { ascending: false });

  const { data, error } = await q.limit(2000);
  if (error) throw new Error(error.message);
  return data || [];
}

export async function getDrill(id) {
  const { data, error } = await supabase
    .from("drills_public")
    .select("*")
    .eq("id", id)
    .single();

  if (error) throw new Error(error.message);
  return data;
}

// ---------- Drills (write = auth) ----------
export async function createDrill(payload) {
  await requireAuth("Login required to add drills.");

  const { data, error } = await supabase
    .from("drills")
    .insert(payload)
    .select("*")
    .single();

  if (error) throw new Error(error.message);
  return data;
}

export async function updateDrill(id, payload) {
  await requireAuth("Login required to edit drills.");

  const { data, error } = await supabase
    .from("drills")
    .update(payload)
    .eq("id", id)
    .select("*")
    .single();

  if (error) throw new Error(error.message);
  return data;
}

export async function deleteDrill(id) {
  await requireAuth("Login required to delete drills.");

  const { error } = await supabase.from("drills").delete().eq("id", id);
  if (error) throw new Error(error.message);
  return { ok: true };
}

// ---------- Comments (public read, insert auth) ----------
export async function listComments(drillId) {
  const { data, error } = await supabase
    .from("drill_comments")
    .select("*")
    .eq("drill_id", drillId)
    .order("created_at", { ascending: false });

  if (error) throw new Error(error.message);
  return data || [];
}

export async function addComment(drillId, payload) {
  await requireAuth("Login required to add comments.");

  const insertPayload = {
    drill_id: drillId,
    comment: payload.comment,
    created_by_name: payload.created_by_name || null,
    created_by_email: payload.created_by_email || null,
  };

  const { data, error } = await supabase
    .from("drill_comments")
    .insert(insertPayload)
    .select("*")
    .single();

  if (error) throw new Error(error.message);
  return data;
}

export async function deleteComment() {
  throw new Error("Deleting comments is disabled.");
}

// ---------- Ratings (public read, insert auth) ----------
export async function listRatings(drillId) {
  const { data, error } = await supabase
    .from("drill_ratings")
    .select("*")
    .eq("drill_id", drillId)
    .order("created_at", { ascending: false });

  if (error) throw new Error(error.message);
  return data || [];
}

export async function addRating(drillId, payload) {
  await requireAuth("Login required to add ratings.");

  const insertPayload = {
    drill_id: drillId,
    score: payload.score,
    created_by_name: payload.created_by_name || null,
    created_by_email: payload.created_by_email || null,
  };

  const { data, error } = await supabase
    .from("drill_ratings")
    .insert(insertPayload)
    .select("*")
    .single();

  if (error) throw new Error(error.message);
  return data;
}

export async function deleteRating() {
  throw new Error("Deleting ratings is disabled.");
}